kiwi/lib/search.py

```diff
@@ -0,0 +1,655 @@
+                'Exploring parameters with '
+                f'{"multivariate " if multivariate else ""}TPE sampler'
+            )
+            sampler = optuna.samplers.TPESampler(
+                seed=config.seed, multivariate=multivariate,
+            )
+        logger.info('Initializing study...')
+        pruner = optuna.pruners.MedianPruner()
+        study = optuna.create_study(
+            study_name=config.search_name,
+            direction='maximize',
+            pruner=pruner,
+            sampler=sampler,
+        )
+
+    # Optimize the study
+    optimize_kwargs = {}
+    if use_mlflow:
+        optimize_kwargs['callbacks'] = [optuna.integration.MLflowCallback()]
+    try:
+        logger.info('Optimizing study...')
+        objective = Objective(config=config, base_config_dict=base_config_dict)
+        study.optimize(
+            objective, n_trials=config.num_trials, **optimize_kwargs,
+        )
+    except KeyboardInterrupt:
+        logger.info('Early stopping search caused by ctrl-C')
+    except Exception as e:
+        logger.error(
+            f'Error occured during search: {e}; '
+            f'current best params are: {study.best_params}'
+        )
+
+    logger.info(f"Saving study to: {output_dir / 'study.pkl'}")
+    joblib.dump(study, output_dir / 'study.pkl')
+
+    try:
+        logger.info('Number of finished trials: {}'.format(len(study.trials)))
+        logger.info('Best trial:')
+        trial = study.best_trial
+        logger.info('  Value: {}'.format(trial.value))
+        logger.info('  Params: ')
+        for key, value in trial.params.items():
+            logger.info('    {}: {}'.format(key, value))
+    except Exception as e:
+        logger.error(f'Logging at end of search failed: {e}')
+
+    if config.num_models_to_keep > 0 and config.num_trials > config.num_models_to_keep:
+        # Keep only n best model checkpoints and remove the rest to free up space
+        try:
+            best_model_paths = objective.best_model_paths
+            models_to_keep = best_model_paths[: config.num_models_to_keep]
+            models_to_remove = best_model_paths[config.num_models_to_keep :]
+            logger.info(
+                f'Keeping the {config.num_models_to_keep} best models '
+                'from the search and removing the other ones.'
+            )
+            logger.info('Models that are kept:')
+            for path in models_to_keep:
+                logger.info(f'\t{path}')
+            for path in models_to_remove:
+                Path(path).unlink()
+        except Exception as e:
+            logger.error(f'Logging at end of search failed: {e}')
+
+    logger.info(f'Saving Optuna plots for this search to: {output_dir}')
+    save_config_to_file(config, output_dir / 'search_config.yaml')
+    try:
+        fig = optuna.visualization.plot_optimization_history(study)
+        fig.write_html(str(output_dir / 'optimization_history.html'))
+    except Exception as e:
```

Comment:
Again but less egregious (has a specific error message). 

I really think you should remove all of these try/except unless there's a very good reason for them. This is not a very good programming practice. Else the entire search can be surrounded by a try/except and always runs successfully!